core/query/mixingQuery.ml

```diff
@@ -0,0 +1,1551 @@
+    | `Bool b -> Constant (Constant.Bool b)
+    | `Char c -> Constant (Constant.Char c)
+    | `Float f -> Constant (Constant.Float f)
+    | `Int i -> Constant (Constant.Int i)
+    | `String s -> Constant (Constant.String s)
+    | `Record fields ->
+        let fields =
+          fields
+          |> List.map (fun (k, v) -> (k, expression_of_base_value v))
+          |> StringMap.from_alist in
+        Record fields
+    | other ->
+        raise (internal_error ("expression_of_base_value undefined for " ^
+          Value.string_of_value other))
+
+  let rec freshen_for_bindings : Var.var Env.Int.t -> t -> t =
+    fun env v ->
+      let ffb = freshen_for_bindings env in
+      let rec var_lookup x =
+        match Env.Int.find_opt x env with
+        | None -> x
+        | Some y -> var_lookup y
+      in
+      match v with
+      | For (tag, gs, os, b) ->
+        let gs', env' =
+          List.fold_left
+            (fun (gs', env') (x, source) ->
+              let y = Var.fresh_raw_var () in
+              ((y, ffb source)::gs', Env.Int.bind x y env'))
+            ([], env)
+            gs
+        in
+          For (tag, List.rev gs', List.map (freshen_for_bindings env') os, freshen_for_bindings env' b)
+      | If (c, t, e) -> If (ffb c, ffb t, ffb e)
+      | Table _ as t -> t
+      | Singleton v -> Singleton (ffb v)
+      | Database db -> Database db
+      | Concat vs -> Concat (List.map ffb vs)
+      | Dedup q -> ffb q
+      | Prom q -> ffb q
+      | Record fields -> Record (StringMap.map ffb fields)
+      | Variant (name, v) -> Variant (name, ffb v)
+      | XML xmlitem -> XML xmlitem
+      | Project (v, name) -> Project (ffb v, name)
+      | Erase (v, names) -> Erase (ffb v, names)
+      | Apply (u, vs) -> Apply (ffb u, List.map ffb vs)
+      | Closure _ as c ->
+        (* we don't attempt to freshen closure bindings *)
+        c
+      | Case (u, cl, d) -> Case (ffb u, StringMap.map (fun (x,y) -> (x, ffb y)) cl, opt_app (fun (x,y) -> Some (x, ffb y)) None d)
+      | Primitive f -> Primitive f
+      | Var (x, ts) as _v ->
+        (* begin
+          match Env.Int.find_opt x env with
+          | None -> v (* Var (x, ts) *)
+          | Some y -> Var (y, ts)
+        end *)
+          Var (var_lookup x, ts)
+      | Constant c -> Constant c
+
+  let flatfield f1 f2 = f1 ^ "@" ^ f2
+
+  let rec flattened_pair x y =
+    match x, y with
+    | Var (_nx, Types.Record row), _ ->
+        let x' = Record (StringMap.fold (fun f _ acc -> StringMap.add f (Project (x,f)) acc) (field_types_of_row row) StringMap.empty)
+        in flattened_pair x' y
+    | _, Var (_ny, Types.Record row) ->
+        let y' = Record (StringMap.fold (fun f _ acc -> StringMap.add f (Project (y,f)) acc) (field_types_of_row row) StringMap.empty)
+        in flattened_pair x y'
+    (* XXX: using a field with an empty name to deal with variables of non-record type ... will it work? *)
```

Comment:
Is this comment still relevant?

Reply:
I'll remove the "will it work" part, as it is tested now. But it makes sense to document the fact that we are using an empty field name for his corner case.